p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,350 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"crypto"
+	"encoding/hex"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/udpmux"
+
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multibase"
+	"github.com/multiformats/go-multihash"
+
+	"github.com/pion/ice/v2"
+	pionlogger "github.com/pion/logging"
+	"github.com/pion/webrtc/v3"
+)
+
+var _ tpt.Listener = &listener{}
+
+const (
+	candidateSetupTimeout         = 20 * time.Second
+	DefaultMaxInFlightConnections = 10
+)
+
+type candidateAddr struct {
+	ufrag string
+	raddr *net.UDPAddr
+}
+
+type listener struct {
+	transport *WebRTCTransport
+
+	mux ice.UDPMux
+
+	config                    webrtc.Configuration
+	localFingerprint          webrtc.DTLSFingerprint
+	localFingerprintMultibase string
+
+	localAddr      net.Addr
+	localMultiaddr ma.Multiaddr
+
+	// buffered incoming connections
+	acceptQueue chan tpt.CapableConn
+
+	// Accepting a connection requires instantiating a peerconnection
+	// and a noise connection which is expensive. We therefore limit
+	// the number of in-flight connection requests. A connection
+	// is considered to be in flight from the instant it is handled
+	// until it is dequed by a call to Accept, or errors out in some
+	// way.
+	inFlightInputQueue     chan candidateAddr
+	maxInFlightConnections uint32
+
+	// used to control the lifecycle of the listener
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+func newListener(transport *WebRTCTransport, laddr ma.Multiaddr, socket net.PacketConn, config webrtc.Configuration) (*listener, error) {
+	localFingerprints, err := config.Certificates[0].GetFingerprints()
+	if err != nil {
+		return nil, err
+	}
+
+	localMh, err := hex.DecodeString(strings.ReplaceAll(localFingerprints[0].Value, ":", ""))
+	if err != nil {
+		return nil, err
+	}
+	localMhBuf, err := multihash.Encode(localMh, multihash.SHA2_256)
+	if err != nil {
+		return nil, err
+	}
+	localFpMultibase, err := multibase.Encode(multibase.Base64url, localMhBuf)
+	if err != nil {
+		return nil, err
+	}
+
+	inFlightQueueCh := make(chan candidateAddr, 1)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	mux := udpmux.NewUDPMux(socket, func(ufrag string, addr net.Addr) bool {
+		// Push to the inFlightQueue asynchronously to avoid blocking the mux goroutine
+		// on candidates being processed. This can cause new connections to fail at high
+		// throughput but will allow packets for existing connections to be processed.
+		select {
+		case inFlightQueueCh <- candidateAddr{ufrag: ufrag, raddr: addr.(*net.UDPAddr)}:
```

Comment:
Should `inFlightQueueCh` have a capacity of larger than 1? Otherwise we'll not be able to handle two connection attempts that arrive at the same time.

Reply:
Why not? One might be blocked for tiny bit, but it will work, no?
Goal of this pipeline is to have us be able to limit how many candidates are in queue (max 10) due to technical limitations of Pion. Yes it is true that the inFlightQueue has a capacity of just 1, but it is not a long blocking one, as it will just accept the conn and then handle it over to the accept queue which has a longer buffer size.

Reply:
There's a `default` statement here. No blocking will occur, not even for a tiny bit. We'll just drop the second connection.

Reply:
Going to revisit this today. Based on this and your previous reviews I seem to get that you:

- want to make sure we block for a bit in case our limit has been reached
- want to make sure not to have always open go routines for it

So will see what to do about this.

Any inspiration @ckousik ?

Reply:
This has now been greatly improved, with a new design. Hopefully this time to your liking.